File: backend/internal/application/game_service.go
package application

import (
	"fmt"
	"math/rand/v2"

	"hotseat/backend/internal/domain"
)

// GameService builds shuffled decks and records finished sessions.
type GameService struct {
	words    domain.WordRepository
	sessions domain.SessionRepository
	rng      *rand.Rand
}

func NewGameService(words domain.WordRepository, sessions domain.SessionRepository, rng *rand.Rand) *GameService {
	return &GameService{words: words, sessions: sessions, rng: rng}
}

// NewDeck builds a deck from the pool of cards that have not appeared yet: a
// card shown in any game stays out of every following deck until the whole
// pool has been played, then the pool resets automatically. The deck is
// shuffled exactly once with Fisher–Yates, so order stays random and playing
// front to back guarantees zero repeats within a session too.
func (g *GameService) NewDeck(difficulty domain.Difficulty) ([]domain.Word, error) {
	if !difficulty.Valid() {
		return nil, fmt.Errorf("invalid difficulty %q", difficulty)
	}
	pool, err := g.words.FindUnseenEnabledByDifficulty(difficulty)
	if err != nil {
		return nil, fmt.Errorf("loading words: %w", err)
	}
	if len(pool) == 0 {
		// Every card has appeared — return them all to the pool and start over.
		if err := g.words.ResetSeen(difficulty); err != nil {
			return nil, fmt.Errorf("resetting card pool: %w", err)
		}
		pool, err = g.words.FindUnseenEnabledByDifficulty(difficulty)
		if err != nil {
			return nil, fmt.Errorf("loading words after reset: %w", err)
		}
	}
	if len(pool) == 0 {
		return nil, fmt.Errorf("no enabled words for difficulty %q", difficulty)
	}
	return FisherYatesShuffle(pool, g.rng), nil
}

// MarkCardAppeared records that a card was shown, keeping it out of future
// decks until the pool resets.
func (g *GameService) MarkCardAppeared(id int64) error {
	return g.words.MarkSeen(id)
}

// ResetCardPool manually returns every card of the difficulty (or all cards
// when difficulty is empty) to the pool.
func (g *GameService) ResetCardPool(difficulty domain.Difficulty) error {
	return g.words.ResetSeen(difficulty)
}

// RecordResult persists a finished session for statistics.
func (g *GameService) RecordResult(result domain.GameSessionResult) error {
	return g.sessions.Insert(result)
}

File: frontend/src/application/gameService.ts
import type { GameSession, GameSessionResult, GameStats } from '@/domain/gameSession';
import type { Team } from '@/domain/team';
import type { Difficulty } from '@/domain/word';
import { fisherYatesShuffle } from '@/lib/shuffle';
import type { SessionRepository } from '@/infrastructure/repositories/sessionRepository';
import type { WordRepository } from '@/infrastructure/repositories/wordRepository';

export class GameService {
  constructor(
    private readonly words: WordRepository,
    private readonly sessions: SessionRepository,
  ) {}

  /**
   * Creates a new session from the pool of cards that have not appeared yet:
   * a card shown in any game stays out of every following deck until the
   * whole pool has been played, at which point the pool resets automatically.
   * The deck is shuffled exactly once (Fisher–Yates), so order stays random
   * and cards never repeat within a session either. The game runs against a
   * countdown of durationMs.
   */
  createSession(difficulty: Difficulty, durationMs: number, team: Team): GameSession {
    if (!Number.isFinite(durationMs) || durationMs <= 0) {
      throw new Error(`Invalid game duration ${durationMs}ms`);
    }
    let pool = this.words.findUnseenEnabledByDifficulty(difficulty);
    if (pool.length === 0) {
      // Every card has appeared — return them all to the pool and start over.
      this.words.resetSeen(difficulty);
      pool = this.words.findUnseenEnabledByDifficulty(difficulty);
    }
    if (pool.length === 0) {
      throw new Error(`No enabled words available for difficulty "${difficulty}"`);
    }
    return {
      id: crypto.randomUUID(),
      difficulty,
      deck: fisherYatesShuffle(pool),
      startedAt: Date.now(),
      durationMs,
      team,
    };
  }

  /** Records that a card was shown, removing it from future decks until the pool resets. */
  markCardAppeared(wordId: number): void {
    this.words.markSeen(wordId);
  }

  /** Manually returns every card to the pool (one difficulty, or all). */
  resetCardPool(difficulty?: Difficulty): void {
    this.words.resetSeen(difficulty);
  }

  /** Persists a finished session for statistics. */
  recordResult(result: GameSessionResult): void {
    this.sessions.insert(result);
  }

  /** Discards a recorded game — it disappears from every leaderboard and stat. */
  discardResult(sessionId: string): void {
    this.sessions.deleteById(sessionId);
  }

  getStats(): GameStats {
    return this.sessions.getStats();
  }
}
